### PR TITLE
[WIP] Multiperiod SWRO flowsheet

### PR DESCRIPTION
**Pull request by arao53**
## Summary/Motivation:
A code example of a similar approach by [DISPATCHES](https://github.com/gmlc-dispatches/dispatches/blob/e5dbfbfbe4e0dd63881c8195af9e489b5ba8ad27/dispatches/models/fossil_case/ultra_supercritical_plant/storage/pricetaker_with_multiperiod_integrated_storage_usc.py)

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py [hunk 1]
```diff
@@ -45,32 +45,43 @@
 from watertap.costing import WaterTAPCosting
 
 
-def main():
+class VariableEfficiency(StrEnum):
+    none = "none"
+    flow = "flow"
+    flow_head = "flow_head"
```
Comment (adam-a-a) [inline watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py]: Couldn't we import this from the modified pump model?

File: watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py [hunk 2]
```diff
@@ -82,9 +93,27 @@ def build():
...
+            }
+        )
+        m.fs.P2 = Pump(
+            default={
+                "property_package": m.fs.properties,
+                "variable_efficiency": VariableEfficiency.flow,
+            }
+        )
+    else:
+        raise NotImplementedError("Variable efficiency type not recognized")
```
Comment (adam-a-a) [inline watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py]:
```suggestion
  
       
      m.fs.P1 = Pump(
          default={
              "property_package": m.fs.properties,
              "variable_efficiency": variable_efficiency,
          }
      )
      m.fs.P2 = Pump(
          default={
              "property_package": m.fs.properties,
              "variable_efficiency": variable_efficiency,
          }
      )
```
I think you can just pass the variable_efficiency arg into the pump config option, and error handling should/could be caught by the pump unit itself.

File: watertap/examples/flowsheets/RO_with_energy_recovery/multiperiod_RO.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+    max_flowrate = 1.5  # kg/s
+    min_flowrate = 0.5  # kg/s
+
+    print("Creating RO flowsheet and MP concrete model...\n")
+
+    # call main to create and initialize model with flow-type variable efficiency
+    m = ConcreteModel()
+    m.ro_mp = swro.main(swro.VariableEfficiency.flow)
+
+    print("\nUnfixing design variables and setting bounds...")
```
Comment (adam-a-a) [inline watertap/examples/flowsheets/RO_with_energy_recovery/multiperiod_RO.py]: Unfixing or fixing?

**Review by adam-a-a — COMMENTED** [review]
Just a quick initial scan- I still have to go through the rest, but looks good so far!

**Comment by lbianchi-lbl** [comment]
@arao53 @adam-a-a To test this PR with IDAES as if IDAES/idaes-pse#908 was already merged, you can set the `idaes-pse` requirement in `setup.py` on line 46 to `idaes-pse @ https://github.com/IDAES/idaes-pse/archive/refs/pull/908/merge.zip`.

**Comment by arao53** [comment]
@adam-a-a @lbianchi-lbl This flowsheet works as expected with the tag for [IDAES/idaes-pse#908](https://github.com/IDAES/idaes-pse/pull/908). This was done with commit d89fe42327687fda3542a6ba77d7eb99ef91c058.

File: watertap/examples/flowsheets/RO_multiperiod_model/pricetaker_multiperiod_RO.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+    m.obj = Objective(
+        expr=(1100 * 0.1 / 365 + 0.1) + sum([blk.electricity_cost for blk in t_blocks]),
+        doc="Daily cost to produce water",
+    )
+
+    m.permeate_yield = Constraint(
+        expr=abs(sum([blk.water_prod for blk in t_blocks]) / daily_water_production - 1)
+        <= 1e-5,
+        doc="The water production over the day is fixed within a tolerance",
+    )
```
Comment (bknueven) [inline watertap/examples/flowsheets/RO_multiperiod_model/pricetaker_multiperiod_RO.py]:
```suggestion
    m.permeate_yield = Constraint(
        expr=(daily_water_production -1e-5,
            sum(blk.water_prod for blk in t_blocks),
            daily_water_production + 1e-5),
        doc="The water production over the day is fixed within a tolerance",
    )
```
I'm surprised Pyomo ingests the `abs`, but in any case this is a nice linear constraint so it isn't necessary.

File: watertap/examples/flowsheets/RO_multiperiod_model/pricetaker_multiperiod_RO.py [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+            sum([blk.weighted_permeate_quality for blk in t_blocks])
+            <= 0.0005 * sum([blk.water_prod for blk in t_blocks])
+        ),
+        doc="Flow-averaged permeate quality must be less than 500 ppm",
+    )
+
+    m.permeate_quality_lb = Constraint(
+        expr=(sum([blk.weighted_permeate_quality for blk in t_blocks]) >= 0),
+        doc="Flow-averaged permeate quality must be greater than 0 ppm",
+    )
```
Comment (bknueven) [inline watertap/examples/flowsheets/RO_multiperiod_model/pricetaker_multiperiod_RO.py]: Any reason not to enforce these block-by-block as opposed to at the aggregate level?

I guess if it goes into one large tank it's okay, but for Ipopt these constraints would be less troublesome enforced at every time step, I think.

File: watertap/examples/flowsheets/RO_multiperiod_model/multiperiod_RO.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+    # Create a dynamic block to store dynamic operation parameters 
+    m.fs.dynamic = Block()
+    m.fs.dynamic.ramp_time = Param(initialize=60,
+                                   units= pyunits.s, 
+                                   mutable=True,
+                                   doc="Time associated with ramping up or down in system pressure")
+    m.fs.dynamic.ramping_rate = Param(initialize=0.7e5,
+                                        units= pyunits.Pa/pyunits.s,
+                                        mutable=True,
+                                        doc="Slowest rate at which pressure can be ramped up or down")
```
Comment (adam-a-a) [inline watertap/examples/flowsheets/RO_multiperiod_model/multiperiod_RO.py:64]:
```suggestion
                                        doc="Fastest rate at which pressure can be ramped up or down")
```

**Comment by adam-a-a** [comment]
Closing this PR as it is outdated, and we will revamp this later in a new PR to provide core functionality for flexible desal.

**Pull request closed without merging**